src/IECore/MeshAlgoDistributePoints.cpp

```diff
@@ -0,0 +1,224 @@
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of Image Engine Design nor the names of any
+//       other contributors to this software may be used to endorse or
+//       promote products derived from this software without specific prior
+//       written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "tbb/tbb.h"
+
+#include "IECore/MeshAlgo.h"
+#include "IECore/MeshPrimitiveEvaluator.h"
+#include "IECore/PointDistribution.h"
+#include "IECore/TriangulateOp.h"
+#include "IECore/TriangleAlgo.h"
+
+using namespace Imath;
+using namespace IECore;
+
+//////////////////////////////////////////////////////////////////////////
+// Distribute Points
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+MeshPrimitivePtr processMesh( const IECore::MeshPrimitive *mesh, const std::string &densityMask, const std::string &uvSet, const std::string &position )
+{
+	if( !mesh )
+	{
+		throw InvalidArgumentException( "MeshAlgo::distributePoints : The input mesh is not valid" );
+	}
+
+	const V3fVectorData *positions = mesh->variableData<const V3fVectorData>( position );
+	if ( !positions )
+	{
+		std::string e = boost::str( boost::format( "MeshAlgo::distributePoints : MeshPrimitive has no suitable \"%s\" primitive variable." ) % position );
+		throw InvalidArgumentException( e );
+	}
+
+	TriangulateOpPtr op = new TriangulateOp();
+	op->inputParameter()->setValue( const_cast<MeshPrimitive *>( mesh ) );
+	op->throwExceptionsParameter()->setTypedValue( false );
+	MeshPrimitivePtr result = runTimeCast<MeshPrimitive>( op->operate() );
+	if ( !result || !result->arePrimitiveVariablesValid() )
+	{
+		throw InvalidArgumentException( "MeshAlgo::distributePoints : The input mesh could not be triangulated" );
+	}
+
+	if( result->variables.find( "faceArea" ) == result->variables.end() )
```

Comment:
Is this use-pre-existing-primvar-if-available behaviour new? I think the Op used to always calculate it, right, rather than trust that the variable means what it thinks it means? I think that generally MeshAlgo functions should be told pretty explicitly what data to operate on rather than make assumptions about naming and validity of data. Other than the obvious P, N etc, we're mostly treating primvars as just being along for the ride, and it being up to the user in Gaffer or wherever to track their meaning/relevance. So here I reckon that'd mean always calculating the areas, or optionally allowing them to be passed in if that's an important optimisation for some known scenarios.

Maybe there's a bigger conversation to be had about general conventions for MeshAlgo, and how it's wrapped in Gaffer? @donboie, any thoughts?

Reply:
Yep, wasn't sure on this one. I made it an optional check to avoid redundant work if, for example, a Gaffer user piped a FaceArea node followed by a Seeds node.

I did pretty intentionally remove the faceAreaPrimVarName options when porting the op though. All those options are just confusion. Nobody ever renames that stuff.

Reply:
> Yep, wasn't sure on this one. I made it an optional check to avoid redundant work if, for example, a Gaffer user piped a FaceArea node followed by a Seeds node.

I don't think primvars can be relied on in that way - they're totally arbitrary, and it's down to the user to track their meaning. The mesh could have been deformed since it was calculated, for instance.

> I did pretty intentionally remove the faceAreaPrimVarName options when porting the op though. All those options are just confusion. Nobody ever renames that stuff.

Not sure I follow here. The names don't matter when porting to MeshAlgo, because we're returning PrimitiveVariables rather than adding them to a mesh. In a hypothetical FaceArea node in Gaffer, I would definitely expect there to be a plug to specify the name of the primvar(s) to create, even if it was most often left at the default value. We talked this through when designing the MeshTangents node - the rationale is as follows :

- The presence of the primvar name plugs and their values makes the node more self documenting - "these input variables will be used to generate these output variables".
- The primvar name plugs also allow users more flexibility if they need it - for instance, to generate face areas for two different uv sets without juggling a whole load of intermediate RenamePrimitiveVariable nodes.
- The default values encourage a convention, which makes things easier to understand for users

I definitely want this nailed down as standard when it comes to designing nodes. I'm less concerned about the signatures of the MeshAlgo functions themselves, but I do think we should discuss the merits of forcing the error handling for missing variables onto the caller - that would currently be my preference. @donboie?

Reply:
So should I move this back to forcing re-calculation of face areas on the grounds the user might have deformed the mesh and invalidated those?

Reply:
Yes, I definitely think so - that matches the previous behaviour and our current assumptions about the genericity of primitive variables.

It might be interesting to think about mechanisms for storing additional information with some specific primitive variables, so they can be relied upon to have specific semantics. For instance, we could store the hash of the data used to generate them, and then use that to know if they a) represent what we think they do and b) are up to date. I think that's outside the scope of this PR though - maybe something to prototype in Grizzly?

Reply:
Addressed in b839c9ff93e34867fb492cfbcbb1f76a69ac1f48